src/decoding/error.rs

```diff
@@ -0,0 +1,119 @@
+    /// Error in the bencode structure (e.g. a missing field end separator).
+    #[fail(display = "bencode encoding corrupted")]
+    StructureError(#[fail(cause)] StructureError),
+    /// Error that occurs if the serialized structure contains an unexpected field.
+    #[fail(display = "unexpected field: {}", _0)]
+    UnexpectedField(String),
+    /// Error through an unexpected bencode token during deserialization.
+    #[fail(display = "discovered {} but expected {}", _0, _1)]
+    UnexpectedToken(String, String),
+}
+
+pub trait ResultExt {
+    fn context(self, context: impl std::fmt::Display) -> Self;
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match &self.context {
+            Some(context) => write!(f, "Error: {} in {}", self.error, context),
+            None => write!(f, "Error: {}", self.error),
+        }
+    }
+}
+
+impl From<ErrorKind> for Error {
+    fn from(error: ErrorKind) -> Self {
+        Error {
+            context: None,
+            error,
+        }
+    }
+}
+
+impl Error {
+    pub fn context(mut self, context: impl Display) -> Self {
+        self.context = Some(context.to_string());
+        self
+    }
+
+    /// Raised when there is a general error while deserializing a type.
+    /// The message should not be capitalized and show not end with a period.
+    pub fn malformed_content(cause: impl Into<failure::Error>) -> Error {
+        Self::from(ErrorKind::MalformedContent(cause.into()))
+    }
+
+    /// Returns a `Error::MissingField` which contains the name of the field.
+    pub fn missing_field(field_name: impl Display) -> Error {
+        Self::from(ErrorKind::MissingField(field_name.to_string()))
+    }
+
+    /// Returns a `Error::UnexpectedField` which contains the name of the field.
+    pub fn unexpected_field(field_name: impl Display) -> Error {
+        Self::from(ErrorKind::UnexpectedField(field_name.to_string()))
+    }
+
+    /// Returns a `Error::UnexpectedElement` which contains a custom error message.
+    pub fn unexpected_token(expected: impl Display, discovered: impl Display) -> Error {
+        Self::from(ErrorKind::UnexpectedToken(
+            expected.to_string(),
+            discovered.to_string(),
+        ))
+    }
+}
+
+impl From<StructureError> for Error {
+    fn from(error: StructureError) -> Self {
+        Self::from(ErrorKind::StructureError(error))
+    }
+}
+
+impl From<num::ParseIntError> for Error {
+    fn from(error: num::ParseIntError) -> Self {
+        Self::malformed_content(error)
+    }
+}
+
+impl From<str::Utf8Error> for Error {
+    fn from(error: str::Utf8Error) -> Self {
+        Self::malformed_content(error)
+    }
+}
+
+impl From<string::FromUtf8Error> for Error {
+    fn from(error: string::FromUtf8Error) -> Self {
+        Self::malformed_content(error)
+    }
+}
+
+impl<T> ResultExt for Result<T, Error> {
+    fn context(self, context: impl std::fmt::Display) -> Result<T, Error> {
```

Comment:
I like this pattern a lot, however, should this add to the context instead of replacing it? My thought is that errors should be able to have their errors incrementally added on to as the decode stack unwinds so that you always get the correct path.

Alternatively, I could see a great benefit to having a separate path field for exactly that purpose

Reply:
Fixed. Context is now concatenated path.